### ISSUE TITLE
[BUG] Adding 5th stepper in UART mode crashes firmware

### ISSUE THREAD

**Issue by RoP09** [bigtreetech/BIGTREETECH-SKR-mini-E3#678]
### Description

I Added a 5th stepper motor driver (TMC2209 V2.0) to my SKR Mini E3 V3 using the  Marlin based FW from [Mrisoc](https://github.com/mriscoc/Ender3V2S1) .

Configuring the driver in Standalone mode works fine, but using UART mode crashes the FW after flashing resulting in an endless loop.

### Steps to reproduce

Configured like below:
<img width="511" alt="image" src="https://user-images.githubusercontent.com/82153569/187460764-6994c2b4-add5-487d-b325-d6c122c3ae8a.png">

**Expected behavior**
TMC2209 works in UART mode

**Actual behavior**
After starting up the printer the FW ends up in endless loop

### Additional Information
This happens regardless if the driver is connected or not. Non UART (Standalone) mode using the same pins works fine.

**Comment by radek8** [comment]
But PD0 and PD2 pins are not RX2 and TX2.
It would be more appropriate to use
PD2 //RX5
PD3 //TX5

![image](https://user-images.githubusercontent.com/46979052/187794069-6f85d3d7-9d80-41f3-8788-a75a46d320ef.png)

**Comment by RoP09** [comment]
> But PD0 and PD2 pins are not RX2 and TX2. It would be more appropriate to use PD2 //RX5 PD3 //TX5
> 
Thanks for the reply! The comments are wrong...
Actually I was initially trying RX2/TX2 from the TFT port on the board (since they are "real" UART pins) but that lead to the same behaviour (board reset).

But I will try with these pins as well!

**Comment by RoP09** [comment]
Same thing. Also results in endless loop of the FW if I connect it to PD2/PD3...

**Comment by RoP09** [comment]
Update: Connected it still same as above to PD0 PD2 PD3 PD4 and PD5 now (for enable pin), and set the config differently now to mserial5 like below:

<img width="365" alt="image" src="https://user-images.githubusercontent.com/82153569/188328147-ceb9e45d-9d6b-4ccc-b204-65bb3668ef2d.png">

Now the board boots up, but then displays TMC CONNECTION ERROR on the display. The printer works fine for a couple of seconds, I can move the other steppers or home it etc. but without the 2nd Z Stepper and after some seconds it reboots again.

Something wrong with my connection?

Connected MS1 and MS2 to 3.3V, bridged both PDN pins and connected them to the RX/TX pins with 1K on TX and the rest like above. Anything I am missing?

**Comment by RoP09** [comment]
Update: I now soldered also the small solder pad to enable UART on the Pin 5 like described in the thread linked in the previous post and also here https://de.aliexpress.com/item/33025720969.html?af=701906&gatewayAdapt=glo2deu.

Now I can even control the stepper from the printer menu even if the TMC connectino error is still displayed. But it seems like it is only half the speed of the other stepper now in this configuration?

**Comment by radek8** [comment]
Did you use the same wiring as the other controllers?

![image](https://user-images.githubusercontent.com/46979052/188332305-6da80c18-96f6-4649-9d61-bca4e1447b32.png)

**Comment by radek8** [comment]
Did you enable it in the configuration.h file?
#define Z2_DRIVER_TYPE TMC2209

**Comment by RoP09** [comment]
> Did you enable it in the configuration.h file? #define Z2_DRIVER_TYPE TMC2209

Yes exactly. If I put it as _STANDALONE, the driver works fine but not in UART mode then (with same wiring).

**Comment by RoP09** [comment]
> Did you use the same wiring as the other controllers?
> 
> ![image](https://user-images.githubusercontent.com/46979052/188332481-518d12f3-c862-41e8-8c13-92c7f9a7f462.png)

Not exactly. I have no 100 Ohms before TX/RX (only the 1K resistor I have). Also DIAG pin I have not connected and CLK + PDN_UART are also not connected to GND (did not read anywhere that this has to be done).

**Comment by radek8** [comment]
> 
> Not exactly. I have no 100 Ohms before TX/RX (only the 1K resistor I have). Also DIAG pin I have not connected and CLK + PDN_UART are also not connected to GND (did not read anywhere that this has to be done).

So that's okay.

**Comment by radek8** [comment]
Something is wrong here.
Active parameters should not be grayed out

![image](https://user-images.githubusercontent.com/46979052/188431818-6154ce6d-f664-4a68-b03f-a9103d8144ef.png)

**Comment by RoP09** [comment]
I am not sure, think that was just a display issue of VS code here, cause normally if I don't define Z2_SERIAL_TX/RX_PIN it will give me compilation errors. If I define Z2_HARDWARE_SERIAL it seems it does not need those RX/TX pin defines and compiles fine.

**Comment by radek8** [comment]
Have you also tried swapping the RX and TX pins?
Sometimes the problem is that the wires to the pins are reversed.
Does not connect TX to TX.
RX connects to TX and TX connects to RX. The one who transmits must connect to the one who listens and vice versa.

Next, I would try to use TX2 and RX2 and on the TFT connector and define it as MSerial2

**Comment by RoP09** [comment]
No did not try swapping them, but I will try that and also give it a try with MSerial2 later!
Thanks for the hints!

**Comment by RoP09** [comment]
Swapped PD2 and PD3 defines now without using MSerial, but only the RX / TX defines - same story (FW freezes, as written in the initial post).
Next I will try to exchange the actualy RX / TX cables and use MSerial5 again...

**Comment by RoP09** [comment]
No success. Also swapping RX TX physical pins results in same error as before...

**Comment by radek8** [comment]
If it doesn't work for you in UART mode, you can remap E0 as Z2 and use the external controller for the extruder in STANDALONE mode.
It's not ideal, but it's usable.

**Comment by RoP09** [comment]
Thanks for the idea, at least E0 I need for LIN_ADVANCE in UART mode, but maybe another stepper (X or Y) could be remapped indeed...
